m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -524,58 +519,75 @@ public Thread newThread(Runnable r) {
 
         @Override
         public void run() {
-            while (!isShutdown.get()) {
-                try {
-                    // Check whether flush has been requested by the reporter
-                    if (shouldFlush.compareAndSet(true, false)) {
-                        flushBuffered();
-                    }
+            boolean semaphoreAcquired = true;
+            try {
+                processorsSemaphore.acquire();
+            } catch (InterruptedException e) {
+                // We may get interrupted if the reporter is closed, in which case
+                // we proceed with shutdown tasks.
+                semaphoreAcquired = false;
```

Comment:
Maybe more clear to immediately shutdown and return here?

Reply:
Agreed.